File: compat/suites/cli/internal/harness/harness.go
// Package harness provides the core test framework for the Overcast compat CLI suite.
//
// Tests emit NDJSON events to stdout. Each test calls the AWS CLI via exec.Command
// and fails if the command exits non-zero or produces unexpected output.
package harness

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// TestFn is the signature for a single test.
type TestFn func(ctx context.Context, t *TestContext) error

// TestCase describes one test within a group.
type TestCase struct {
	Name string
	Fn   TestFn
	Op   string
	Skip string
	// NA, if non-empty, causes the test to be emitted with status "na".
	// Use this when the AWS CLI does not yet expose this operation.
	// NA results are excluded from pass-rate calculations.
	NA string
	// Depends lists test names in the same group that must pass before this
	// test runs.  If any dependency failed or was skipped, this test is
	// automatically skipped.
	Depends []string
}

// TestGroup is a collection of related tests with optional setup/teardown.
type TestGroup struct {
	Suite    string
	Service  string
	Name     string
	Tests    []TestCase
	Setup    func(ctx context.Context, t *TestContext) error
	Teardown func(ctx context.Context, t *TestContext) error
}

// TestContext carries per-run state for tests.
type TestContext struct {
	Endpoint string
	Region   string
	RunID    string

	mu    sync.Mutex
	state map[string]any
}

// NewTestContext creates a fresh TestContext.
func NewTestContext(endpoint, region, runID string) *TestContext {
	return &TestContext{
		Endpoint: endpoint,
		Region:   region,
		RunID:    runID,
		state:    make(map[string]any),
	}
}

// Set stores a value in the context state bag.
func (t *TestContext) Set(key string, val any) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.state[key] = val
}

// Get retrieves a value from the context state bag.
func (t *TestContext) Get(key string) (any, bool) {
	t.mu.Lock()
	defer t.mu.Unlock()
	v, ok := t.state[key]
	return v, ok
}

// GetString retrieves a string value from the state bag.
func (t *TestContext) GetString(key string) string {
	v, ok := t.Get(key)
	if !ok {
		return ""
	}
	s, _ := v.(string)
	return s
}

// Log writes a debug message to stderr.
func (t *TestContext) Log(msg string) {
	fmt.Fprintln(os.Stderr, "[cli]", msg)
}

// Reset clears the state bag (called between groups).
func (t *TestContext) Reset() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.state = make(map[string]any)
}

// IsUnimplemented reports whether an error represents a 501 / not-implemented response.
func IsUnimplemented(err error) bool {
	if err == nil {
		return false
	}
	s := err.Error()
	return strings.Contains(s, "501") ||
		strings.Contains(s, "NotImplemented") ||
		strings.Contains(s, "UnknownOperationException") ||
		strings.Contains(s, "UnsupportedOperation") ||
		strings.Contains(s, "not implemented")
}

// ─── NDJSON events ────────────────────────────────────────────────────────────

// emitMu serialises writes to stdout so concurrent goroutines (parallel
// group execution) never interleave partial NDJSON lines.
var emitMu sync.Mutex

func emit(obj any) {
	b, _ := json.Marshal(obj)
	emitMu.Lock()
	os.Stdout.Write(b)
	os.Stdout.Write([]byte("\n"))
	emitMu.Unlock()
}

// GroupCounts holds the result totals from a single RunGroup call.
type GroupCounts struct {
	Passed, Failed, Skipped, Unimplemented, Cancelled int
}

// RunGroup executes a single TestGroup, emitting one test_result per test.
// It returns the aggregate counts for the caller to roll up into run_end.
func RunGroup(ctx context.Context, g TestGroup) GroupCounts {
	t := NewTestContext("", "", "")

	// Extract endpoint/region/runID from context values if present.
	if v, ok := ctx.Value(ctxEndpoint{}).(string); ok {
		t.Endpoint = v
	}
	if v, ok := ctx.Value(ctxRegion{}).(string); ok {
		t.Region = v
	}
	if v, ok := ctx.Value(ctxRunID{}).(string); ok {
		t.RunID = v
	}

	var counts GroupCounts
	failedOrSkipped := map[string]bool{}

	if g.Setup != nil {
		if err := g.Setup(ctx, t); err != nil {
			emit(map[string]any{"event": "group_setup_error", "suite": g.Suite, "group": g.Name, "error": err.Error()})
			for _, tc := range g.Tests {
				emit(map[string]any{
					"event": "test_result", "suite": g.Suite, "service": g.Service,
					"group": g.Name, "test": tc.Name, "status": "skip",
					"error": fmt.Sprintf("setup failed: %v", err), "duration_ms": 0,
				})
				counts.Skipped++
			}
			return counts
		}
	}

	for _, tc := range g.Tests {
		if ctx.Err() != nil {
			emit(cancelledEvent{Event: "cancelled", Suite: g.Suite, Group: g.Name, Test: tc.Name})
			counts.Cancelled++
			continue
		}
		if tc.NA != "" {
			emit(map[string]any{
				"event": "test_result", "suite": g.Suite, "service": g.Service,
				"group": g.Name, "test": tc.Name, "status": "na",
				"error": tc.NA, "duration_ms": 0,
			})
			// na is excluded from pass-rate counters
			continue
		}
		if tc.Skip != "" {
			emit(map[string]any{
				"event": "test_result", "suite": g.Suite, "service": g.Service,
				"group": g.Name, "test": tc.Name, "status": "skip",
				"error": tc.Skip, "duration_ms": 0,
			})
			counts.Skipped++
			failedOrSkipped[tc.Name] = true
			continue
		}

		// Dependency gate — skip if any declared dependency failed or was skipped.
		if len(tc.Depends) > 0 {
			var failedDeps []string
			for _, dep := range tc.Depends {
				if failedOrSkipped[dep] {
					failedDeps = append(failedDeps, dep)
				}
			}
			if len(failedDeps) > 0 {
				emit(map[string]any{
					"event": "test_result", "suite": g.Suite, "service": g.Service,
					"group": g.Name, "test": tc.Name, "status": "skip",
					"error":       fmt.Sprintf("dependency failed: %s", strings.Join(failedDeps, ", ")),
					"duration_ms": 0,
				})
				counts.Skipped++
				failedOrSkipped[tc.Name] = true
				continue
			}
		}

		start := time.Now()
		err := tc.Fn(ctx, t)
		durMs := time.Since(start).Milliseconds()

		if err == nil {
			emit(map[string]any{
				"event": "test_result", "suite": g.Suite, "service": g.Service,
				"group": g.Name, "test": tc.Name, "status": "pass", "duration_ms": durMs,
			})
			counts.Passed++
		} else if IsUnimplemented(err) {
			emit(map[string]any{
				"event": "test_result", "suite": g.Suite, "service": g.Service,
				"group": g.Name, "test": tc.Name, "status": "unimplemented", "duration_ms": durMs,
			})
			counts.Unimplemented++
			failedOrSkipped[tc.Name] = true
		} else {
			emit(map[string]any{
				"event": "test_result", "suite": g.Suite, "service": g.Service,
				"group": g.Name, "test": tc.Name, "status": "fail",
				"error": err.Error(), "duration_ms": durMs,
			})
			counts.Failed++
			failedOrSkipped[tc.Name] = true
		}
	}

	if g.Teardown != nil {
		if err := g.Teardown(ctx, t); err != nil {
			emit(map[string]any{"event": "group_teardown_error", "suite": g.Suite, "group": g.Name, "error": err.Error()})
		}
	}
	return counts
}

// RunSuite executes all groups in parallel.
// RunGroup already creates its own TestContext per group, so no shared state
// is at risk from concurrent execution.
func RunSuite(suite string, groups []TestGroup, endpoint, region, runID string) {
	ctx := context.WithValue(context.Background(), ctxEndpoint{}, endpoint)
	ctx = context.WithValue(ctx, ctxRegion{}, region)
	ctx = context.WithValue(ctx, ctxRunID{}, runID)

	total := 0
	for _, g := range groups {
		total += len(g.Tests)
	}
	startedAt := time.Now()
	emit(map[string]any{
		"event": "run_start", "suite": suite, "run_id": runID,
		"total_tests": total, "endpoint": endpoint,
		"started_at": startedAt.UTC().Format(time.RFC3339),
	})

	// Limit concurrent group execution. OVERCAST_COMPAT_PARALLEL_SLOTS is
	// injected by the Go runner; default 8.
	slots := 8
	if v := os.Getenv("OVERCAST_COMPAT_PARALLEL_SLOTS"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			slots = n
		}
	}
	sem := make(chan struct{}, slots)

	// Pre-allocate one slot per group so goroutines can write without locking.
	groupResults := make([]GroupCounts, len(groups))
	var wg sync.WaitGroup
	for i, g := range groups {
		wg.Add(1)
		go func(i int, g TestGroup) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			// Per-group timeout: prevents a single hung CLI call from
			// blocking this semaphore slot (and thus the whole suite) forever.
			groupCtx, groupCancel := context.WithTimeout(ctx, 5*time.Minute)
			defer groupCancel()
			groupResults[i] = RunGroup(groupCtx, g)
		}(i, g)
	}
	wg.Wait()

	var passed, failed, skipped, unimplemented int
	for _, c := range groupResults {
		passed += c.Passed
		failed += c.Failed
		skipped += c.Skipped
		unimplemented += c.Unimplemented
	}
	emit(map[string]any{
		"event": "run_end", "suite": suite, "run_id": runID,
		"passed": passed, "failed": failed, "skipped": skipped,
		"unimplemented": unimplemented,
		"duration_ms":   time.Since(startedAt).Milliseconds(),
	})
}

// context key types.
type ctxEndpoint struct{}
type ctxRegion struct{}
type ctxRunID struct{}

// NewRunContext returns a context with endpoint, region, and runID values
// that RunGroup will extract.
func NewRunContext(ctx context.Context, endpoint, region, runID string) context.Context {
	ctx = context.WithValue(ctx, ctxEndpoint{}, endpoint)
	ctx = context.WithValue(ctx, ctxRegion{}, region)
	ctx = context.WithValue(ctx, ctxRunID{}, runID)
	return ctx
}

// ── Interactive mode types ────────────────────────────────────────────────

// StdinCommand represents a command received on stdin.
type StdinCommand struct {
	Command string    `json:"command"`
	BatchID string    `json:"batch_id,omitempty"`
	Tests   []TestRef `json:"tests,omitempty"`
	Group   string    `json:"group,omitempty"`
	Test    string    `json:"test,omitempty"`
}

// TestRef identifies a group and optional subset of tests.
type TestRef struct {
	Group string   `json:"group"`
	Tests []string `json:"tests,omitempty"`
}

type buildingEvent struct {
	Event   string `json:"event"`
	Suite   string `json:"suite"`
	Message string `json:"message"`
}

type readyEvent struct {
	Event      string `json:"event"`
	Suite      string `json:"suite"`
	TotalTests int    `json:"total_tests"`
}

type batchCompleteEvent struct {
	Event         string `json:"event"`
	Suite         string `json:"suite"`
	BatchID       string `json:"batch_id"`
	Passed        int    `json:"passed"`
	Failed        int    `json:"failed"`
	Skipped       int    `json:"skipped"`
	Unimplemented int    `json:"unimplemented"`
	Cancelled     int    `json:"cancelled"`
	DurationMs    int64  `json:"duration_ms"`
}

type cancelledEvent struct {
	Event   string `json:"event"`
	Suite   string `json:"suite"`
	BatchID string `json:"batch_id,omitempty"`
	Group   string `json:"group,omitempty"`
	Test    string `json:"test,omitempty"`
	Reason  string `json:"reason,omitempty"`
}

// EmitBuilding emits a building event during interactive setup.
func EmitBuilding(suite, message string) {
	emit(buildingEvent{Event: "building", Suite: suite, Message: message})
}

// EmitReady emits a ready event when the suite is prepared for commands.
func EmitReady(suite string, totalTests int) {
	emit(readyEvent{Event: "ready", Suite: suite, TotalTests: totalTests})
}

// EmitBatchComplete emits a batch_complete event after a batch finishes.
func EmitBatchComplete(suite, batchID string, totals GroupCounts, durationMs int64) {
	emit(batchCompleteEvent{
		Event: "batch_complete", Suite: suite, BatchID: batchID,
		Passed: totals.Passed, Failed: totals.Failed,
		Skipped: totals.Skipped, Unimplemented: totals.Unimplemented,
		Cancelled: totals.Cancelled, DurationMs: durationMs,
	})
}

// EmitPong responds to an orchestrator ping with the currently executing test.
func EmitPong(suite, runningTest string) {
	emit(struct {
		Event       string `json:"event"`
		Suite       string `json:"suite"`
		RunningTest string `json:"running_test"`
	}{Event: "pong", Suite: suite, RunningTest: runningTest})
}

// ReadCommands reads NDJSON commands from stdin and sends them to the returned channel.
// It closes the channel when stdin is closed.
func ReadCommands() <-chan StdinCommand {
	ch := make(chan StdinCommand, 16)
	go func() {
		defer close(ch)
		scanner := bufio.NewScanner(os.Stdin)
		scanner.Buffer(make([]byte, 64*1024), 1024*1024)
		for scanner.Scan() {
			line := scanner.Text()
			if strings.TrimSpace(line) == "" {
				continue
			}
			var cmd StdinCommand
			if err := json.Unmarshal([]byte(line), &cmd); err != nil {
				fmt.Fprintf(os.Stderr, "[harness] invalid JSON on stdin: %s\n", line)
				continue
			}
			ch <- cmd
		}
	}()
	return ch
}

File: compat/suites/cdk/src/lib/harness.ts
export interface TestContext {
  endpoint: string;
  region: string;
  runId: string;
  stackName: string;
  log(msg: string): void;
  signal?: AbortSignal;
  [key: string]: unknown;
}

export interface SuiteContext {
  endpoint: string;
  region: string;
  runId: string;
  stackName: string;
}

export type TestFn = (ctx: TestContext) => Promise<void>;

export interface TestCase {
  name: string;
  fn: TestFn;
  skip?: boolean | string;
  depends?: string[];
}

export interface TestGroup {
  suite: string;
  service: string;
  name: string;
  tests: TestCase[];
  setup?: (ctx: TestContext) => Promise<void>;
  teardown?: (ctx: TestContext) => Promise<void>;
}

interface RunStartEvent {
  event: "run_start";
  suite: string;
  started_at: string;
  endpoint: string;
  version: "1";
  total_tests?: number;
}

interface TestStartEvent {
  event: "test_start";
  suite: string;
  service: string;
  group: string;
  test: string;
}

interface TestResultEvent {
  event: "test_result";
  suite: string;
  service: string;
  group: string;
  test: string;
  status: "pass" | "fail" | "skip" | "unimplemented" | "cancelled";
  duration_ms: number;
  error?: string;
}

interface RunEndEvent {
  event: "run_end";
  suite: string;
  passed: number;
  failed: number;
  skipped: number;
  unimplemented: number;
  duration_ms: number;
}

// ─── Interactive-mode event shapes ────────────────────────────────────────

interface BuildingEvent {
  event: "building";
  suite: string;
  message: string;
}

interface ReadyEvent {
  event: "ready";
  suite: string;
  total_tests: number;
}

interface BatchCompleteEvent {
  event: "batch_complete";
  suite: string;
  batch_id: string;
  passed: number;
  failed: number;
  skipped: number;
  unimplemented: number;
  cancelled: number;
  duration_ms: number;
}

interface PongEvent {
  event: "pong";
  suite: string;
  running_test: string;
}

export function emitEvent(
  ev:
    | RunStartEvent
    | TestStartEvent
    | TestResultEvent
    | RunEndEvent
    | BuildingEvent
    | ReadyEvent
    | BatchCompleteEvent
    | PongEvent,
): void {
  process.stdout.write(JSON.stringify(ev) + "\n");
}

const TEST_TIMEOUT_MS = 5 * 60_000;

function withTimeout<T>(promise: Promise<T>, label: string): Promise<T> {
  let timer!: ReturnType<typeof setTimeout>;
  const timeout = new Promise<never>((_, reject) => {
    timer = setTimeout(
      () =>
        reject(
          new Error(`${label} timed out after ${TEST_TIMEOUT_MS / 1000}s`),
        ),
      TEST_TIMEOUT_MS,
    );
  });
  return Promise.race([promise, timeout]).finally(() =>
    clearTimeout(timer),
  ) as Promise<T>;
}

// ─── Abort helpers ────────────────────────────────────────────────────────

/** Return true if the error represents an AbortController cancellation. */
function isAbortError(err: unknown): boolean {
  if (err instanceof DOMException && err.name === "AbortError") return true;
  if (
    err != null &&
    typeof err === "object" &&
    (err as Record<string, unknown>).name === "AbortError"
  )
    return true;
  return false;
}

/**
 * Race a promise against an AbortSignal. Rejects with AbortError if the
 * signal fires before the promise settles.
 */
function raceAbort<T>(promise: Promise<T>, signal: AbortSignal): Promise<T> {
  if (signal.aborted) {
    return Promise.reject(new DOMException("Cancelled", "AbortError"));
  }
  return new Promise<T>((resolve, reject) => {
    const onAbort = () => reject(new DOMException("Cancelled", "AbortError"));
    signal.addEventListener("abort", onAbort, { once: true });
    promise.then(
      (val) => {
        signal.removeEventListener("abort", onAbort);
        resolve(val);
      },
      (err) => {
        signal.removeEventListener("abort", onAbort);
        reject(err);
      },
    );
  });
}

export function isUnimplemented(err: unknown): boolean {
  if (err == null || typeof err !== "object") return false;
  const e = err as Record<string, unknown>;
  const msg = String(e["message"] ?? "");

  const meta = e["$metadata"];
  if (meta != null && typeof meta === "object") {
    if ((meta as Record<string, unknown>)["httpStatusCode"] === 501)
      return true;
  }

  if (e["name"] === "UnknownOperationException") return true;
  if (e["name"] === "NotImplemented") return true;

  return (
    msg.includes("HTTP 501") ||
    msg.includes("NotImplemented") ||
    msg.includes("UnknownOperationException") ||
    msg.includes("x-emulator-unsupported")
  );
}

export interface RunGroupOptions {
  /** Map of "group:test" → AbortController for cancellation support. */
  abortControllers?: Map<string, AbortController>;
  /** Batch ID for event correlation in interactive mode. */
  batchId?: string;
}

export async function runGroup(
  group: TestGroup,
  ctx: TestContext,
  options?: RunGroupOptions,
): Promise<{
  passed: number;
  failed: number;
  skipped: number;
  unimplemented: number;
  cancelled: number;
}> {
  let passed = 0;
  let failed = 0;
  let skipped = 0;
  let unimplemented = 0;
  let cancelled = 0;
  const failedOrSkipped = new Set<string>();

  if (group.setup) {
    try {
      await withTimeout(group.setup(ctx), `setup ${group.name}`);
    } catch (err) {
      const reason = `setup failed: ${String(err)}`;
      for (const tc of group.tests) {
        emitEvent({
          event: "test_result",
          suite: group.suite,
          service: group.service,
          group: group.name,
          test: tc.name,
          status: "skip",
          duration_ms: 0,
          error: reason,
        });
        skipped++;
      }
      ctx.log(`[${group.name}] ${reason}`);
      return { passed, failed, skipped, unimplemented, cancelled };
    }
  }

  try {
    for (const tc of group.tests) {
      if (tc.skip) {
        emitEvent({
          event: "test_result",
          suite: group.suite,
          service: group.service,
          group: group.name,
          test: tc.name,
          status: "skip",
          duration_ms: 0,
          error: typeof tc.skip === "string" ? tc.skip : undefined,
        });
        skipped++;
        failedOrSkipped.add(tc.name);
        continue;
      }

      if (tc.depends && tc.depends.length > 0) {
        const failedDeps = tc.depends.filter((dep) => failedOrSkipped.has(dep));
        if (failedDeps.length > 0) {
          emitEvent({
            event: "test_result",
            suite: group.suite,
            service: group.service,
            group: group.name,
            test: tc.name,
            status: "skip",
            duration_ms: 0,
            error: `dependency failed: ${failedDeps.join(", ")}`,
          });
          skipped++;
          failedOrSkipped.add(tc.name);
          continue;
        }
      }

      // Create per-test AbortController for cancellation support.
      const ac = new AbortController();
      const acKey = `${group.name}:${tc.name}`;
      options?.abortControllers?.set(acKey, ac);

      ctx.signal = ac.signal;

      emitEvent({
        event: "test_start",
        suite: group.suite,
        service: group.service,
        group: group.name,
        test: tc.name,
      });

      const started = Date.now();
      try {
        await withTimeout(raceAbort(tc.fn(ctx), ac.signal), tc.name);
        emitEvent({
          event: "test_result",
          suite: group.suite,
          service: group.service,
          group: group.name,
          test: tc.name,
          status: "pass",
          duration_ms: Date.now() - started,
        });
        passed++;
      } catch (err) {
        const error =
          err instanceof Error ? `${err.name}: ${err.message}` : String(err);
        if (isAbortError(err)) {
          emitEvent({
            event: "test_result",
            suite: group.suite,
            service: group.service,
            group: group.name,
            test: tc.name,
            status: "cancelled",
            duration_ms: Date.now() - started,
            error: "cancelled",
          });
          cancelled++;
          failedOrSkipped.add(tc.name);
        } else if (isUnimplemented(err)) {
          emitEvent({
            event: "test_result",
            suite: group.suite,
            service: group.service,
            group: group.name,
            test: tc.name,
            status: "unimplemented",
            duration_ms: Date.now() - started,
            error,
          });
          unimplemented++;
          failedOrSkipped.add(tc.name);
        } else {
          emitEvent({
            event: "test_result",
            suite: group.suite,
            service: group.service,
            group: group.name,
            test: tc.name,
            status: "fail",
            duration_ms: Date.now() - started,
            error,
          });
          failed++;
          failedOrSkipped.add(tc.name);
        }
      } finally {
        options?.abortControllers?.delete(acKey);
      }
    }
  } finally {
    if (group.teardown) {
      try {
        await withTimeout(group.teardown(ctx), `teardown ${group.name}`);
      } catch (err) {
        ctx.log(`[${group.name}] teardown error: ${String(err)}`);
      }
    }
  }

  return { passed, failed, skipped, unimplemented, cancelled };
}

export async function runSuite(
  suite: string,
  groups: TestGroup[],
  baseCtx: SuiteContext,
): Promise<void> {
  const suiteStart = Date.now();
  const total = groups.reduce((sum, g) => sum + g.tests.length, 0);

  emitEvent({
    event: "run_start",
    suite,
    started_at: new Date().toISOString(),
    endpoint: baseCtx.endpoint,
    version: "1",
    total_tests: total,
  });

  const log = (msg: string): void => {
    process.stderr.write(`[compat:${suite}] ${msg}\n`);
  };

  let passed = 0;
  let failed = 0;
  let skipped = 0;
  let unimplemented = 0;

  for (const group of groups) {
    const groupCtx: TestContext = { ...baseCtx, log };
    const r = await runGroup(group, groupCtx);
    passed += r.passed;
    failed += r.failed;
    skipped += r.skipped;
    unimplemented += r.unimplemented;
  }

  emitEvent({
    event: "run_end",
    suite,
    passed,
    failed,
    skipped,
    unimplemented,
    duration_ms: Date.now() - suiteStart,
  });
}

export function makeRunId(): string {
  return (
    "oc-" +
    Math.floor(Math.random() * 0xffffffff)
      .toString(16)
      .padStart(8, "0")
  );
}
